src/join/join.jl

```diff
@@ -461,142 +511,334 @@ end
         else
             _fill_oncols_left_table_left!(_res, _columns(dsl)[j], ranges, new_ends, total_length, missing; threads = threads)
         end
         push!(res, _res)
-
     end
+    
     if dsl isa SubDataset
         newds = Dataset(res, copy(index(dsl)), copycols = false)
     else
         newds = Dataset(res, Index(copy(index(dsl).lookup), copy(index(dsl).names), copy(index(dsl).format)), copycols = false)
     end
 
-
     for j in 1:length(right_cols)
-        _res = allocatecol(_columns(dsr)[right_cols[j]], total_length)
+        if dsr isa SubDataset
+            _res = allocatecol(_columns(copy(dsr))[right_cols[j]], total_length, addmissing = true)
+        else
+            _res = allocatecol(_columns(dsr)[right_cols[j]], total_length, addmissing = true)
+        end
         if DataAPI.refpool(_res) !== nothing
             fill_val = DataAPI.invrefpool(_res)[missing]
-            _fill_right_cols_table_left!(_res.refs, view(DataAPI.refarray(_columns(dsr)[right_cols[j]]), idx), ranges, new_ends, total_length, fill_val; threads = threads)
+            _fill_right_cols_table_left!(_res.refs, view(DataAPI.refarray(_columns(dsr)[right_cols[j]]), idx), ranges, new_ends, total_length, fill_val; ra = ra, inbits = inbits, en2 = revised_ends, threads = threads)
+            # Solve the #undef problem. There has to be a better way
+            for i in 1:length(_res)
+                if !isassigned(_res,i)
+                    _res[i] = missing
+                end
+            end
         else
-            _fill_right_cols_table_left!(_res, view(_columns(dsr)[right_cols[j]], idx), ranges, new_ends, total_length, missing; threads = threads)
+            _fill_right_cols_table_left!(_res, view(DataAPI.refarray(_columns(dsr)[right_cols[j]]), idx), ranges, new_ends, total_length, missing; ra = ra, inbits = inbits, en2 = revised_ends, threads = threads)
         end
         push!(_columns(newds), _res)
-
         new_var_name = make_unique([_names(dsl); _names(dsr)[right_cols[j]]], makeunique = makeunique)[end]
         push!(index(newds), new_var_name)
         setformat!(newds, index(newds)[new_var_name], getformat(dsr, _names(dsr)[right_cols[j]]))
     end
+    
     if multiple_match
         insertcols!(newds, ncol(newds)+1, multiple_match_name => multiple_match_col, unsupported_copy_cols = false, makeunique = makeunique)
     end
+    
     if obs_id[1]
         obs_id_name1 = Symbol(obs_id_name, "_left")
         obs_id_left = allocatecol(nrow(dsl) < typemax(Int32) ? Int32 : Int64, total_length)
         _fill_oncols_left_table_left!(obs_id_left, 1:nrow(dsl), ranges, new_ends, total_length, missing; threads = threads)
-        insertcols!(newds, ncol(newds)+1, obs_id_name1 => obs_id_left, unsupported_copy_cols = false, makeunique = makeunique)
+        insertcols!(dsnewdsl, ncol(newds)+1, obs_id_name1 => obs_id_left, unsupported_copy_cols = false, makeunique = makeunique)
```

Comment:
why `dsnewdsl`? I mean what is the problem with `newds`

Reply:
Sorry, this shoud be newds.